nifi-python-extensions/nifi-text-embeddings-module/src/main/python/vectorstores/QueryQdrant.py

```diff
@@ -0,0 +1,192 @@
+        expression_language_scope=ExpressionLanguageScope.FLOWFILE_ATTRIBUTES,
+    )
+    NUMBER_OF_RESULTS = PropertyDescriptor(
+        name="Number of Results",
+        description="The number of results to return from Qdrant.",
+        required=True,
+        validators=[StandardValidators.POSITIVE_INTEGER_VALIDATOR],
+        default_value="10",
+        expression_language_scope=ExpressionLanguageScope.FLOWFILE_ATTRIBUTES,
+    )
+    FILTER = PropertyDescriptor(
+        name="Metadata Filter",
+        description='Optional metadata filter to apply with the query. For example: { "author": "john.doe" }',
+        required=False,
+        validators=[StandardValidators.NON_EMPTY_VALIDATOR],
+        expression_language_scope=ExpressionLanguageScope.FLOWFILE_ATTRIBUTES,
+    )
+
+    properties = (
+        QdrantUtils.QDRANT_PROPERTIES
+        + QdrantUtils.EMBEDDING_MODEL_PROPERTIES
+        + [
+            QUERY,
+            FILTER,
+            NUMBER_OF_RESULTS,
+            QueryUtils.OUTPUT_STRATEGY,
+            QueryUtils.RESULTS_FIELD,
+            QueryUtils.INCLUDE_METADATAS,
+            QueryUtils.INCLUDE_DISTANCES,
+        ]
+    )
+
+    embeddings = None
+    query_utils = None
+    client = None
+
+    def __init__(self, **kwargs):
+        pass
+
+    def getPropertyDescriptors(self):
+        return self.properties
+
+    def onScheduled(self, context):
+        self.client = QdrantClient(
+            url=context.getProperty(QdrantUtils.QDRANT_URL).getValue(),
+            api_key=context.getProperty(QdrantUtils.QDRANT_API_KEY).getValue(),
+            prefer_grpc=context.getProperty(QdrantUtils.PREFER_GRPC).asBoolean(),
+            https=context.getProperty(QdrantUtils.HTTPS).asBoolean(),
+        )
+        self.embeddings = create_embedding_service(context)
+        self.query_utils = QueryUtils.QueryUtils(context)
+
+    def transform(self, context, flowfile):
+        collection_name = (
+            context.getProperty(QdrantUtils.COLLECTION_NAME)
+            .evaluateAttributeExpressions(flowfile)
+            .getValue()
+        )
+        query = (
+            context.getProperty(self.QUERY)
+            .evaluateAttributeExpressions(flowfile)
+            .getValue()
+        )
+        num_results = (
+            context.getProperty(self.NUMBER_OF_RESULTS)
+            .evaluateAttributeExpressions(flowfile)
+            .asInteger()
+        )
+        filter = (
+            context.getProperty(self.FILTER)
+            .evaluateAttributeExpressions(flowfile)
+            .getValue()
+        )
+        vector_store = Qdrant(
+            client=self.client,
+            collection_name=collection_name,
+            embeddings=self.embeddings,
+        )
+        results = vector_store.similarity_search_with_score(
+            query=query,
+            k=num_results,
+            filter=None if filter is None else json.loads(filter),
```

Comment:
I'm far from being an expert in this domain and I never used Qdrant before but wondering if we could take the opportunity of this PR to also consider support for hybrid search and provide an easy way for a user to search based on semantic (embeddings) but also keywords. Do we have examples on how to do that with Qdrant? how could hybrid search be leveraged?

Reply:
Qdrant has support for dense and sparse embeddings.
For hybrid search, a user has to perform 2 searches(sparse and dense) and use a custom fusion strategy for combining the scores.

Reply:
OK, thanks, that's useful information. I was hoping for some sort of higher level API making it easier for the user :)

Reply:
The Python client library does have abstractions like the `query()`  method, which handles sparse and dense embedding generation, fusion etc. But that's something done on the client side.

Reply:
Yeah this is the point I was trying to make: it would be nice to have a processor that makes it super easy for a user in NiFi to do the hybrid search without the need to do in a multi-steps process. For example, have an optional property in the configuration of QueryQdrant for a user to provide keywords (the user may put those as FlowFile attributes upstream in the flow) and use the QueryQdrant processor where it's doing both searches and the fusion of the results.

Reply:
That `query()` method downloads the embedding models locally and run the inference, which could slow down the NiFi deployment.

Reply:
OK, I see, thanks for the details